libsolidity/analysis/TypeChecker.cpp

```diff
@@ -338,35 +340,59 @@ bool TypeChecker::visit(FunctionDefinition const& _function)
 		if (_function.isOrdinary() && !_function.isPartOfExternalInterface())
 			m_errorReporter.typeError(5587_error, _function.location(), "Internal functions cannot be payable.");
 	}
+
+	vector<VariableDeclaration const*> internalParametersInConstructor;
+
 	auto checkArgumentAndReturnParameter = [&](VariableDeclaration const& var) {
 		if (type(var)->containsNestedMapping())
 			if (var.referenceLocation() == VariableDeclaration::Location::Storage)
 				solAssert(
-					_function.libraryFunction() || !_function.isPublic(),
+					_function.libraryFunction() || _function.isConstructor() || !_function.isPublic(),
 					"Mapping types for parameters or return variables "
 					"can only be used in internal or library functions."
 				);
-		if (_function.isPublic())
+		bool functionIsExternallyVisible =
+			(!_function.isConstructor() && _function.isPublic()) ||
+			(_function.isConstructor() && !m_currentContract->abstract());
+		if (
+			_function.isConstructor() &&
+			var.referenceLocation() == VariableDeclaration::Location::Storage
+		)
+			// TODO what about other things like internal function pointers?
+			internalParametersInConstructor.emplace_back(&var);
+
+		if (functionIsExternallyVisible)
 		{
 			auto iType = type(var)->interfaceType(_function.libraryFunction());
 
 			if (!iType)
 			{
-				solAssert(!iType.message().empty(), "Expected detailed error message!");
-				m_errorReporter.typeError(4103_error, var.location(), iType.message());
+				string message = iType.message();
+				solAssert(!message.empty(), "Expected detailed error message!");
+				if (_function.isConstructor())
+					message += " You can make the contract abstract to avoid this problem.";
+				m_errorReporter.typeError(4103_error, var.location(), message);
 			}
 		}
 		if (
-			_function.isPublic() &&
+			functionIsExternallyVisible &&
 			!experimentalFeatureActive(ExperimentalFeature::ABIEncoderV2) &&
 			!typeSupportedByOldABIEncoder(*type(var), _function.libraryFunction())
 		)
+		{
+			string message =
+				"This type is only supported in ABIEncoderV2. "
+				"Use \"pragma experimental ABIEncoderV2;\" to enable the feature.";
+			if (_function.isConstructor())
+				message +=
+					" Alternatively, make the contract abstract and supply the "
```

Comment:
This message doesn't seem to be in any test (at least in the ones modified here)

Reply:
good catch! Will add a test.